src/MongoDB.Driver.Core/Core/Servers/Server.cs

```diff
@@ -343,26 +395,58 @@ private bool IsRecovering(ServerErrorCode code, string message)
             return false;
         }
 
+        private bool IsRecoveringErrorException(Exception exception)
+        {
+            return
+                exception is MongoCommandException commandException &&
+                IsRecovering((ServerErrorCode)commandException.Code, commandException.ErrorMessage);
+        }
+
+        private void SetDescription(ServerDescription newDescription)
+        {
+            var oldDescription = Interlocked.CompareExchange(ref _currentDescription, null, null);
+            SetDescription(oldDescription, newDescription);
+        }
+
+        private void SetDescription(ServerDescription oldDescription, ServerDescription newDescription)
+        {
+            Interlocked.Exchange(ref _currentDescription, newDescription);
+            OnDescriptionChanged(sender: this, new ServerDescriptionChangedEventArgs(oldDescription, newDescription));
+        }
+
         private bool ShouldClearConnectionPoolForChannelException(Exception ex, SemanticVersion serverVersion)
         {
-            if (ex is MongoNotPrimaryException mongoNotPrimaryException && mongoNotPrimaryException.Code == (int)ServerErrorCode.NotMaster)
+            if (ex is MongoConnectionException mongoCommandException &&
+                mongoCommandException.IsNetworkException &&
+                !mongoCommandException.ContainsSocketTimeoutException)
             {
-                return !Feature.KeepConnectionPoolWhenNotMasterConnectionException.IsSupported(serverVersion);
+                return true;
             }
-
-            return true;
+            if (IsNotMasterErrorException(ex) || IsRecoveringErrorException(ex))
+            {
+                return
+                    IsShutdownErrorException(ex) ||
+                    !Feature.KeepConnectionPoolWhenNotMasterConnectionException.IsSupported(serverVersion); // i.e. serverVersion < 4.1.10
```

Comment:
This is the updated logic for [CSHARP-2433](https://jira.mongodb.org/browse/CSHARP-2433) https://github.com/DmitryLukyanov/mongo-csharp-driver/pull/35

As discussed offline:  The  wording for that spec was later clarified as part of this spec PR: https://github.com/mongodb/specifications/pull/730/files, which was described as "no drivers changes" required but perhaps should have been marked as "driver changes MAY be required" . The key thing to note is that https://github.com/mongodb/specifications/pull/730/files clarified that when handling a "not master" or "node is recovering error" that we should clear the connection pool if and only if the the error is "node is shutting down" OR the error originated from server version < 4.2.